core/src/main/scala/kafka/log/AbstractIndex.scala

```diff
@@ -202,8 +202,26 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
    * @throws IOException if rename fails
    */
   def renameTo(f: File) {
-    try Utils.atomicMoveWithFallback(file.toPath, f.toPath)
-    finally file = f
+    maybeLock(lock) {
+      val position = if (this.mmap == null) 0 else this.mmap.position()
+      if (OperatingSystem.IS_WINDOWS && this.mmap != null)
+        forceUnmap()
+      try {
+        Utils.atomicMoveWithFallback(file.toPath, f.toPath)
+        if (OperatingSystem.IS_WINDOWS) {
+          val raf = new RandomAccessFile(f, "rw")
+          try {
+            val len = raf.length()
+            this.mmap = raf.getChannel.map(FileChannel.MapMode.READ_WRITE, 0, len)
+            this.mmap.position(position)
+          } finally {
+            CoreUtils.swallow(raf.close(), this)
+          }
+        }
+      } finally {
+        file = f
+      }
+    }
```

Comment:
This code is pretty confusing. I suggest having a separate path for Windows instead of mixing the logic.